### PR TITLE
Ability to configure devDir with flag [#21]

### PR DESCRIPTION
**Pull request by glosee**
Potential fix for issue #21 that adds ability to use a flag called `--dev-dir`, and a check for `XDG_CACHE_DIR` before `HOME`.

The `--dev-dir` flag is the be-all-end-all result. If this flag is not used, then the application looks for `XDG_CACHE_DIR` first, then `HOME`, then `USERPROFILE` as it used to.

### CODE REVIEW

File: lib/node-gyp.js [hunk 1]
```diff
@@ -98,6 +86,7 @@ proto.configDefs = {
   , 'tarball': String // 'install'
   , jobs: String      // 'build'
   , thin: String      // 'configure'
+  , 'dev-dir': String // everywhere
```
Comment (TooTallNate) [inline lib/node-gyp.js]: Flag name should be `--devdir` to parallel the `--nodedir` flag.

**Comment by dantran** [comment]
@TooTallNate, please merge this PR with your suggested change

**Comment by glosee** [comment]
@TooTallNate I completely forgot about this PR until @dantran commented about making your suggested change. I have updated the code to use `devdir` instead of `dev-dir`. Let me know if there is anything else that should change. Thanks!

**Comment by evaryont** [comment]
A reminder/vote for this PR. Is it OK to merge?

**Comment by saper** [comment]
Very useful concept, but can we name it more meaningfully (we also have `nodedir`) ? This is really a cache for the node configuration and include files, so something with `node` and `cache` would be better: `cachenodeinc` maybe?

**Comment by jasonkarns** [comment]
The default for `XDG_CACHE_HOME` is `$HOME/.cache`, so to truly follow the XDG basedir spec, that should be used if `XDG_CACHE_HOME` isn't set (and before `HOME`)


I'll also point out that under the XDG dir, the node-gyp dir ought to be `node-gyp`, not `.node-gyp`.

So I think the directory priority ought to be:

1. `$XDG_CACHE_HOME/node-gyp`
2. `$HOME/.cache/node-gyp`
3. `$HOME/.node-gyp`
4. `$USERPROFILE/.node-gyp`

**Comment by bnoordhuis** [comment]
Making devDir configurable seems reasonable to me but I'm less convinced that honoring XDG_CACHE_HOME is a good thing.  For one, the way it's done in this PR would force everyone to download tarballs again (because the cache directory changes) when it's released.

I agree with @saper that the name is not great.  Something like `--cachedir` better describes what it does.

Also, it would be good to have a regression test or two.

**Comment by jasonkarns** [comment]
> For one, the way it's done in this PR would force everyone to download tarballs again (because the cache directory changes) when it's released.

That's a valid reason against the way this PR is written, but I don't think it's necessarily a strike against XDG_CACHE_HOME in principle.

An option is to choose the directory based on which one *exists* and not by which env var is set (in the order I [listed](https://github.com/nodejs/node-gyp/pull/540#issuecomment-180057762)). This would prevent re-downloading everything. And it would allow a manual (one-time) opt-in for users to just move their `~/.node-gyp` dir into the XDG location if desired.

If none of the directories already exist, then choose (and create) the directory based on the env vars.

**Comment by bnoordhuis** [comment]
What I missed in the discussion in #21 is why following XDG_CACHE_HOME and XDG in general is a good thing.

XDG is not a thing on Windows - people will instead use `set npm_config_cachedir=c:\temp` or whatever we end up naming the flag - so I'm leaning towards simply not bothering and being consistent across platforms.

**Comment by jasonkarns** [comment]
There may not be XDG_* env vars on windows, but the default XDG cache dir is `$HOME/.cache` so using `$HOME/.cache/node-gyp` would be perfectly consistent across platforms. And leveraging the XDG env vars puts node-gyp in line with a host of other tools that let users manage their directory structure through the XDG env vars.

Put simply, if we agree that allowing users to configure where the node-gyp dir is (and I think that's evident with the assumption that `cachedir` is a useful option), then it logically follows that adhering to a spec that makes it possible for users to configure directory locations *across a large and growing number of apps* with just a couple env vars is vastly superior to app-specific flags/config.

**Comment by bnoordhuis** [comment]
Okay, I would suggest that @glosee starts with just the `--cachedir` option and then you or @glosee can follow up with a XDG compatibility PR.  I haven't made up my mind yet about whether said compatibility is preferable but that doesn't need to hold up the core functionality.

**Comment by jasonkarns** [comment]
sounds good

**Comment by bnoordhuis** [comment]
Closing, superseded by #916 which landed earlier today.

**Comment by jcrben** [comment]
Is there a reason why it seems that https://github.com/nodejs/node-gyp/pull/916 declined to go with:

1. dev-dir flag
2. XDG_CACHE_HOME 
3. fall back to $HOME?

I saw some discussion about Windows but that doesn't seem like a good reason to ignore the Unix platform's preferences...

**Pull request closed without merging**